keras_contrib/layers/normalization.py

```diff
@@ -1,3 +1,235 @@
+
+        if self.initial_weights is not None:
+            self.set_weights(self.initial_weights)
+            del self.initial_weights
+        self.built = True
+
+    def call(self, x, mask=None):
+        if self.mode == 0 or self.mode == 2:
+            assert self.built, 'Layer must be built before being called'
+            input_shape = K.int_shape(x)
+
+            reduction_axes = list(range(len(input_shape)))
+            del reduction_axes[self.axis]
+            broadcast_shape = [1] * len(input_shape)
+            broadcast_shape[self.axis] = input_shape[self.axis]
+
+            if K.backend() == 'tf':
+                mean_batch, var = K.KTF.nn.moments(x, reduction_axes, shift=None, name=None, keep_dims=False)
+                std_batch = (K.sqrt(var + self.epsilon))
+            else:
+                mean_batch = K.mean(x, axis=reduction_axes)
+                std_batch = K.sqrt(K.var(x, axis=reduction_axes) + self.epsilon)
+
+            r_max_val = K.get_value(self.r_max)
+            r = std_batch / self.running_std
+            r = K.stop_gradient(K.clip(r, 1 / r_max_val, r_max_val))
+
+            d_max_val = K.get_value(self.d_max)
+            d = (mean_batch - self.running_mean) / self.running_mean
+            d = K.stop_gradient(K.clip(d, -d_max_val, d_max_val))
+
+            if sorted(reduction_axes) == range(K.ndim(x))[:-1]:
+                x_normed_batch = (x - mean_batch) / std_batch
+                x_normed = (x_normed_batch * r + d) * self.gamma + self.beta
+            else:
+                # need broadcasting
+                broadcast_mean = K.reshape(mean_batch, broadcast_shape)
+                broadcast_std = K.reshape(std_batch, broadcast_shape)
+                broadcast_r = K.reshape(r, broadcast_shape)
+                broadcast_d = K.reshape(d, broadcast_shape)
+                broadcast_beta = K.reshape(self.beta, broadcast_shape)
+                broadcast_gamma = K.reshape(self.gamma, broadcast_shape)
+
+                x_normed_batch = (x - broadcast_mean) / broadcast_std
+                x_normed = (x_normed_batch * broadcast_r + broadcast_d) * broadcast_gamma + broadcast_beta
+
+            # explicit update to moving mean and standard deviation
+            self.add_update([K.moving_average_update(self.running_mean, mean_batch, self.momentum),
+                             K.moving_average_update(self.running_std, std_batch ** 2, self.momentum)], x)
+
+            # update r_max and d_max
+            t_val = K.get_value(self.t)
+            r_val = self.r_max_value / (1 + (self.r_max_value - 1) * np.exp(-t_val))
+            d_val = self.d_max_value / (1 + ((self.d_max_value / 1e-3) - 1) * np.exp(-(2 * t_val)))
+            t_val += 1.
+
+            self.add_update([K.update(self.r_max, r_val),
+                             K.update(self.d_max, d_val),
+                             K.update(self.t, t_val)], x)
+
+            if self.mode == 0:
```

Comment:
No, looking at it again see the inference diagram at the end of the algorithm. It states y = gamma * (x - running_mean) / (running_std) + beta.

There is no mention of r and d here. The results of my testing this algorithm also show similarities to what the paper described. Validation accuracies of both BatchNorm and BatchRenorm should be very similar if all other things are kept constant. Batch renorm may give a very small boost to the final result however, as is the case in the paper when using Inception v3 over ImageNet. 

This also corresponds to my own readings comparing the scores of WRN-16-4 with Batchnorm and Batchrenorm. Both are behaving similarly, with batch renorm giving slightly better performance. The inference portion is important here and cannot be removed, since inference in BRn should be similar to BN inference.